src/Symfony/Component/Security/Http/Authenticator/AccessToken/AccessTokenAuthenticator.php

```diff
@@ -0,0 +1,156 @@
+    private ?string $realm;
+    /**
+     * @var AccessTokenExtractorInterface[]
+     */
+    private iterable $accessTokenExtractors;
+
+    /**
+     * @param AccessTokenExtractorInterface[] $accessTokenExtractors
+     */
+    public function __construct(UserProviderInterface $userProvider, AccessTokenHandlerInterface $accessTokenHandler, iterable $accessTokenExtractors, AuthenticationSuccessHandlerInterface $successHandler = null, AuthenticationFailureHandlerInterface $failureHandler = null, string $realm = null)
+    {
+        $this->successHandler = $successHandler;
+        $this->failureHandler = $failureHandler;
+        $this->userProvider = $userProvider;
+        $this->accessTokenHandler = $accessTokenHandler;
+        $this->accessTokenExtractors = $accessTokenExtractors;
+        $this->realm = $realm;
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return null !== $this->extractAccessToken($request);
+    }
+
+    public function authenticate(Request $request): Passport
+    {
+        $accessToken = $this->extractAccessToken($request);
+        if (null === $accessToken) {
+            throw new BadCredentialsException('Invalid credentials.');
+        }
+        try {
+            $userIdentifier = $this->accessTokenHandler->getUserIdentifierFrom($accessToken);
+        } catch (AuthenticationException $e) {
+            throw $e;
+        } catch (Throwable $e) {
+            throw new BadCredentialsException('Invalid credentials.', 0, $e);
+        }
+
+        return new Passport(
+            new UserBadge($userIdentifier, $this->userProvider->loadUserByIdentifier(...)),
+            new AccessTokenCredentials($accessToken)
+        );
+    }
+
+    public function createToken(Passport $passport, string $firewallName): TokenInterface
+    {
+        return new AccessToken($passport->getUser(), $firewallName, $passport->getUser()->getRoles());
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
+    {
+        if (null === $this->successHandler) {
+            return null; // let the original request continue
+        }
+
+        return $this->successHandler->onAuthenticationSuccess($request, $token);
+    }
+
+    public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
+    {
+        if (null === $this->failureHandler) {
+            if (null !== $this->translator) {
+                $errorMessage = $this->translator->trans($exception->getMessageKey(), $exception->getMessageData(), 'security');
+            } else {
+                $errorMessage = strtr($exception->getMessageKey(), $exception->getMessageData());
+            }
+
+            return new Response(
+                null,
+                Response::HTTP_UNAUTHORIZED,
+                ['WWW-Authenticate' => $this->getAuthenticateHeader($errorMessage)]
+            );
```

Comment:
Could we extract this code into a class that implements `AuthenticationFailureHandlerInterface`? In the constructor, when no failure handler is passed, we could set `$this->failureHandler` to an instance of this new class. This way, we would not need to perform that `null` check anymore.